kotlinx-coroutines-core/jdk8/src/future/Future.kt

```diff
@@ -190,3 +190,18 @@ private class ContinuationHandler<T>(
         }
     }
 }
+
+private class CancelFutureOnCompletion(
+    private val future: Future<*>
+) : JobNode() {
+    override fun invoke(cause: Throwable?) {
+        // Don't interrupt when cancelling future on completion, because no one is going to reset this
+        // interruption flag and it will cause spurious failures elsewhere.
+        // We do not cancel the future if it's already completed in some way,
+        // because `cancel` on a completed future won't change the state but is not guaranteed to behave well
+        // on reentrancy. See https://github.com/Kotlin/kotlinx.coroutines/issues/4156
+        if (cause != null && !future.isDone) future.cancel(false)
```

Comment:
> but is not guaranteed to behave well on reentrancy
Can you please mention what exactly goes wrong? Something like `Reentrant cancellation of a future may cause repeated cancellation of it's dependants`

Reply:
I don't think the documentation says anything specific, unfortunately, so what exactly goes wrong seems to be an implementation detail. In the issue linked, the problem was that `future.cancel` shared the job of calling the remaining `handle` callbacks, which called `future.cancel`, which shared the job, and so on recursively in the same thread.